File: headers_decoding/jiansheku.js
const CryptoJS = require('crypto-js');

f = (new Date).getTime()

// data = {
//     'eid': '',
//     'achievementQueryType': 'and',
//     'achievementQueryDto': [],
//     'personnelQueryDto': {
//         'queryType': 'and',
//     },
//     'aptitudeQueryDto': {
//         'queryType': 'and',
//         'nameStr': '',
//         'aptitudeQueryType': 'and',
//         'businessScopeQueryType': 'or',
//         'filePlaceType': '1',
//         'aptitudeDtoList': [
//             {
//                 'codeStr': '',
//                 'queryType': 'and',
//                 'aptitudeType': 'qualification',
//             },
//         ],
//         'aptitudeSource': 'new',
//     },
//     'page': {
//         'page': 1,
//         'limit': 20,
//         'field': '',
//         'order': '',
//     },
// }

gd = function (e) {
    var t = new Array
        , n = 0;
    for (var i in e)
        t[n] = i,
            n++;
    return t.sort()
}

_d = function e(t) {
    var n;
    if (Array.isArray(t)) {
        for (var r in n = new Array,
            t) {
            var o = t[r];
            for (var i in o)
                null == o[i] ? delete t[r][i] : Array.isArray(t[r][i]) && e(t[r][i])
        }
        return n = t,
            JSON.stringify(n).replace(/^(\s|")+|(\s|")+$/g, "")
    }
    return n = t && t.constructor === Object ? JSON.stringify(t) : t
}

md = function (e) {
    var t = gd(e)
        , n = "";
    for (var i in t) {
        var r = _d(e[t[i]]);
        null != r && "" != r.toString() && (n += t[i] + "=" + r + "&")
    }
    return n
}

pd = function (e, t, time) {
    var n = t + e + time;
    return n = CryptoJS.MD5(n).toString();
}

function getSign(data, f) {
    sign = (param = data,
        time = f,
        t = md(param),
        pd("ghaepVf6IhcHmgnk4NCTXLApxQkBcvh1",
            pd("mwMlWOdyM7OXbjzQPulT1ndRZIAjShDB",
                pd("ZuSj0gwgsKXP4fTEz55oAG2q2p1SVGKK", t, time), time), time))

    return sign
}

console.log(getSign(data, f));

File: request_headers/jiansheku.py
# 网页: https://www.jiansheku.com/search/enterprise/
# API(翻页获得): https://capi.jiansheku.com/nationzj/enterprice/page
# 需要逆向的是请求头中的sign参数

import time

import execjs
import requests

page = input('请输入页码：')

json_data = {
    'eid': '',
    'achievementQueryType': 'and',
    'achievementQueryDto': [],
    'personnelQueryDto': {
        'queryType': 'and',
    },
    'aptitudeQueryDto': {
        'queryType': 'and',
        'nameStr': '',
        'aptitudeQueryType': 'and',
        'businessScopeQueryType': 'or',
        'filePlaceType': '1',
        'aptitudeDtoList': [
            {
                'codeStr': '',
                'queryType': 'and',
                'aptitudeType': 'qualification',
            },
        ],
        'aptitudeSource': 'new',
    },
    'page': {
        'page': page,
        'limit': 20,
        'field': '',
        'order': '',
    },
}

headers = {
    'Accept': 'application/json, text/plain, */*',
    'Accept-Language': 'zh-CN,zh;q=0.9,ja;q=0.8,en;q=0.7,zh-TW;q=0.6',
    'Connection': 'keep-alive',
    'Content-Type': 'application/json;charset=UTF-8',
    'Origin': 'https://www.jiansheku.com',
    'Referer': 'https://www.jiansheku.com/',
    'Sec-Fetch-Dest': 'empty',
    'Sec-Fetch-Mode': 'cors',
    'Sec-Fetch-Site': 'same-site',
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/134.0.0.0 Safari/537.36',
    'deviceType': 'PC',
    'page': 'search-enterprise',
    'sec-ch-ua': '"Chromium";v="134", "Not:A-Brand";v="24", "Google Chrome";v="134"',
    'sec-ch-ua-mobile': '?0',
    'sec-ch-ua-platform': '"Windows"',
    'sign': '7e53f1a2da1d10536b462fce4509eb3c',
    'timestamp': '1741854747559',
}

timestamp = int(round(time.time() * 1000))

with open('jiansheku.js', 'r', encoding='utf-8') as f:
    js_code = f.read()
    ctx = execjs.compile(js_code)

headers['sign'] = ctx.call('getSign', json_data, timestamp)
headers['timestamp'] = str(timestamp)

response = requests.post('https://capi.jiansheku.com/nationzj/enterprice/page', headers=headers, json=json_data)

print(response.json())
